### DISCUSSION TITLE
Pinning pg_partman to older version (4.7.4) in apt

### DISCUSSION THREAD

**Discussion by dtseiler** [pgpartman/pg_partman#587] [General]
I'm in the middle of some big migrations so having to worry about possible breakage from a surprise pg_partman package update was something I'd like to avoid. After discussing with @keithf4, I decided the simples & safest bet would be to just pin the partman package at 4.7.4 for the time being until my other major projects are done and I can focus on the partman 5.0.0 upgrade steps.

These instructions have been tested on Ubuntu 22.04 (jammy) with the PGDG apt repository setup.

First, you need to add the apt-archive PGDG repo, as that is where 4.7.4 and all older version can be found. Typically you would do this by adding these lines to your `/etc/apt/sources.list`:

```
deb https://apt-archive.postgresql.org/pub/repos/apt jammy-pgdg-archive main
deb-src https://apt-archive.postgresql.org/pub/repos/apt jammy-pgdg-archive main
```

Next, pin the desired version by adding a file with these contents to `/etc/apt/preferences.d/` (assumes you are on postgresql-15, change if otherwise):

```
Package: postgresql-15-partman
Pin: version 4.7.4-2.pgdg22.04+1
Pin-Priority: 999
```

You want the priority to be higher than the default (typically `500`). I picked `999` to be sure of that.

Now you should be able to update and install the package and it will install `4.7.4-2` (or keep it there)

```
sudo apt-get update
sudo apt-get install postgresql-15-partman
```

Hope this helps if you're in a similar situation.

**Comment by dtseiler** [comment]
Here is an example of achieving the same thing via `chef`:

```ruby
  # - set up the PGDG apt-archive repo
  apt_repository 'pgdg-archive' do
    uri "https://apt-archive.postgresql.org/pub/repos/apt"
    key "https://www.postgresql.org/media/keys/ACCC4CF8.asc"
    distribution ""
    components ["#{node['lsb']['codename']}-pgdg-archive", 'main']
    action :add
  end

  # - pin the desired version
  apt_preference 'pg_partman' do
    package_name      "postgresql-#{node['postgres']['version']}-partman"
    pin               "version #{node['postgres']['pg_partman']['version']}"
    pin_priority      900
  end

  # Install pg_partman
  package "postgresql-#{node['postgres']['version']}-partman"
```

**Comment by keithf4** [comment]
Thanks for this!
